### PR TITLE
[cdc] Added pulsar database cdc synchronization action

### PR DESCRIPTION
**Pull request by zhuangchong**
### Purpose

Added pulsar database cdc synchronization action and database synchronization base class .

### CODE REVIEW

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/MessageQueueSyncDatabaseActionBase.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    // ------------------------------------------------------------------------
+    //  Flink run methods
+    // ------------------------------------------------------------------------
+
+    @Override
+    public void run() throws Exception {
+        build();
+        execute(String.format("KAFKA-Paimon Database Sync: %s", database));
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/MessageQueueSyncDatabaseActionBase.java:195]: `execute(jobName());`

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/SyncTableActionBase.java [hunk 2]
```diff
@@ -214,17 +211,6 @@ public void build() throws Exception {
...
-        if (source instanceof Source) {
-            return env.fromSource(
-                    (Source<String, ?, ?>) source, WatermarkStrategy.noWatermarks(), sourceName());
-        }
-        if (source instanceof SourceFunction) {
-            return env.addSource((SourceFunction<String>) source, sourceName());
-        }
-        throw new UnsupportedOperationException("Unrecognized source type");
-    }
-
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/SyncTableActionBase.java:227]: Why not keep this method?

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/pulsar/PulsarActionUtils.java [hunk 3]
```diff
@@ -324,8 +324,7 @@ static DataFormat getDataFormat(Configuration pulsarConfig) {
     /** Referenced to {@link PulsarPartitionSplitReader#createPulsarConsumer}. */
     static MessageQueueSchemaUtils.ConsumerWrapper createPulsarConsumer(
             Configuration pulsarConfig, String topic) throws PulsarClientException {
-        SourceConfiguration pulsarSourceConfiguration =
-                toSourceConfiguration(preprocessPulsarConfig(pulsarConfig));
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/pulsar/PulsarActionUtils.java:328]: Why modify this?

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/pulsar/PulsarSyncDatabaseAction.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.Map;
+
+/** Synchronize database from Pulsar. */
+public class PulsarSyncDatabaseAction extends MessageQueueSyncDatabaseActionBase {
+
+    public PulsarSyncDatabaseAction(
+            String warehouse,
+            String database,
+            Map<String, String> catalogConfig,
+            Map<String, String> kafkaConfig) {
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/pulsar/PulsarSyncDatabaseAction.java:36]: Rename `kafkaConfig`.

**Review by yuzelin — COMMENTED** [review]
Thanks for contributing. Left some comments.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/pulsar/PulsarActionUtils.java (hunk 3 above)
Comment (zhuangchong) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/pulsar/PulsarActionUtils.java:328]: I found that the `preprocessPulsarConfig` method will be executed first in the `toSourceConfiguration` method, so preprocessPulsarConfig is deleted here.

```java
    static SourceConfiguration toSourceConfiguration(Configuration rawConfig) {
        return new SourceConfiguration(preprocessPulsarConfig(rawConfig));
    }
```

File: docs/layouts/shortcodes/generated/pulsar_sync_database.html [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        <td><h5>--warehouse</h5></td>
+        <td>The path to Paimon warehouse.</td>
+    </tr>
+    <tr>
+        <td><h5>--database</h5></td>
+        <td>The database name in Paimon catalog.</td>
+    </tr>
+    <tr>
+        <td><h5>--schema-init-max-read</h5></td>
+        <td>If your tables are all from a topic, you can set this parameter to initialize the number of tables to be synchronized. The default value is 1000.</td>
```
[the remark this answers is not in the document]
Comment (zhuangchong) [reply docs/layouts/shortcodes/generated/pulsar_sync_database.html:38]: done.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/MessageQueueSyncDatabaseActionBase.java (hunk 1 above)
Comment (zhuangchong) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/MessageQueueSyncDatabaseActionBase.java:195]: done.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/SyncTableActionBase.java (hunk 2 above)
Comment (zhuangchong) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/SyncTableActionBase.java:227]: I think this part of the code is best written in each implementation class. When a table action is added later, env.addSource() or env.fromSource() or others can be used directly and explicitly.

https://github.com/apache/incubator-paimon/blob/1571f8f33ff74bcc9a8a0b6301da349c1780dae2/paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSyncTableAction.java#L119-L125

Comment (yuzelin) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/SyncTableActionBase.java:227]: I think we can keep this method. We can modify it in the future if needed.

Comment (zhuangchong) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/SyncTableActionBase.java]: done.

**Review by yuzelin — APPROVED** [review]
+1

**Pull request merged**